### PR TITLE
add email_logging to User Structure and User Role Structure

### PR DESCRIPTION
**Pull request by obregos**
Add email_logging to User Structure and User Role Structure
issue #390

### CODE REVIEW

**Comment by jubilee2** [comment]
If we determine this change is from 14.4.0, you can use following code.
```
if(is.null(version) || utils::compareVersion(version, "14.4.0") < 0) NULL else character(0),
```

**Comment by spgarbet** [comment]
@jubilee2 I think that's something we should add to this patch.

**Comment by spgarbet** [comment]
==> devtools::check(args = c('--as-cran','--no-multiarch'))

══ Documenting ══════════════════════════════════════════════════════
ℹ Updating redcapAPI documentation
ℹ Loading redcapAPI
Error in `load_all()`:
! Failed to load R/redcapDataStructure.R
Caused by error in `if (is.na(a)) ...`:
! the condition has length > 1
Backtrace:
     ▆
  1. ├─devtools::check(args = c("--as-cran", "--no-multiarch"), check_dir = dirname(getwd()))
  2. │ └─devtools::document(pkg, quiet = quiet)
  3. │   └─roxygen2::roxygenise(pkg$path, roclets)
  4. │     └─roxygen2 (local) load_code(base_path)
  5. │       └─pkgload::load_all(path, helpers = FALSE, attach_testthat = FALSE)
  6. │         └─pkgload::load_code(path, quiet = quiet)
  7. │           ├─withr (local) withr_with_dir(path, source_many(paths, encoding, env))
  8. │           │ └─base::force(code)
  9. │           └─pkgload:::source_many(paths, encoding, env)
 10. │             ├─rlang::try_fetch(...)
 11. │             │ ├─base::tryCatch(...)
 12. │             │ │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
 13. │             │ │   └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 14. │             │ │     └─base (local) doTryCatch(return(expr), name, parentenv, handler)
 15. │             │ └─base::withCallingHandlers(...)
 16. │             └─pkgload:::source_one(file, encoding, envir = envir)
 17. │               └─base::eval(exprs[i], envir)
 18. │                 └─base::eval(exprs[i], envir)
 19. │                   ├─base::data.frame(...) at redcapAPI/R/redcapDataStructure.R:389:1
 20. │                   └─utils::compareVersion(version, "14.4.0")
 21. └─base::.handleSimpleError(...)
 22.   └─rlang (local) h(simpleError(msg, call))
 23.     └─handlers[[1L]](cnd)
 24.       └─cli::cli_abort(msg, parent = cnd, call = quote(load_all()))
 25.         └─rlang::abort(...)
Execution halted

Exited with status 1.

This structure isn't "functionalized" like the other to pass in version.

**Comment by spgarbet** [comment]
── R CMD check results ───────────────────────── redcapAPI 2.9.2 ────
Duration: 38.7s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔

R CMD check succeeded

File: R/redcapConnection.R [hunk 1]
```diff
@@ -788,7 +788,7 @@ offlineConnection <- function(meta_data = NULL,
                        redcap_data = redcapUserStructure(version))
   this_user_roles <- 
     validateRedcapData(data = .offlineConnection_readFile(user_roles), 
-                       redcap_data = REDCAP_USER_ROLE_STRUCTURE)
+                       redcap_data = redcapUserRoleStructure("14.4.0"))
```
Comment (jubilee2) [inline R/redcapConnection.R:791]: can use `version` to replace "14.4.0"

File: R/redcapConnection.R [hunk 2]
```diff
@@ -901,7 +901,7 @@ offlineConnection <- function(meta_data = NULL,
       has_user_roles = function() !is.null(this_user_roles), 
       flush_user_roles = function() this_user_roles <<- NULL, 
       refresh_user_roles = function(x) {this_user_roles <<- validateRedcapData(data = .offlineConnection_readFile(x), 
-                                                                               redcap_data = REDCAP_USER_ROLE_STRUCTURE)}, 
+                                                                               redcap_data = redcapUserRoleStructure("14.4.0"))},
```
Comment (jubilee2) [inline R/redcapConnection.R:904]: can use `this_version` to replace "14.4.0"

**Review by jubilee2 — COMMENTED** [review]
Here are some suggestion

**Comment by spgarbet** [comment]
── R CMD check results ───────────────────────── redcapAPI 2.9.3 ────
Duration: 45s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔

R CMD check succeeded

File: R/redcapConnection.R (hunk 1 above)
Comment (spgarbet) [reply R/redcapConnection.R]: done

File: R/redcapConnection.R (hunk 2 above)
Comment (spgarbet) [reply R/redcapConnection.R]: done

**Review by jubilee2 — APPROVED** [review]
Looks good to me and passed the test above! 👍

**Comment by spgarbet** [comment]
── R CMD check results ───────────────────────── redcapAPI 2.9.3 ────
Duration: 40.6s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔

R CMD check succeeded

**Comment by spgarbet** [comment]
── R CMD check results ─────────────────────────────── redcapAPI 2.9.3 ────
Duration: 39s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔

R CMD check succeeded

**Comment by spgarbet** [comment]
```
── Failed tests ───────────────────────────────────────────────────────────
Failure (test-205-exportTypedRecords-offline-Functionality.R:147:5): Invalid data is reported, links not generated without the version number
`exportRecordsTyped(rcon_off, cast = list(system = as.character))` produced unexpected warnings.
Expected match: Version number not stored
Actual values:
* Some records failed validation. Use `reviewInvalidRecords` to review the failures.

Failure (test-205-exportTypedRecords-offline-Functionality.R:153:5): Invalid data is reported, links not generated without the version number
Inv$link_to_form not equal to c(NA_character_, NA_character_).
2/2 mismatches
x[1]: "https://redcap.vumc.org/redcap_v14.4.0/DataEntry/index.php?pid=18842
x[1]: 5&page=branching_logic&id=18&event_id=513748"
y[1]: NA

x[2]: "https://redcap.vumc.org/redcap_v14.4.0/DataEntry/index.php?pid=18842
x[2]: 5&page=branching_logic&id=19&event_id=513748"
y[2]: NA

[ FAIL 2 | WARN 0 | SKIP 10 | PASS 2063 ]
```

**Comment by spgarbet** [comment]
```
══ Results ════════════════════════════════════════════════════════════════
Duration: 412.1 s

── Skipped tests (10) ─────────────────────────────────────────────────────
• No logs for test window in past (1):
  test-800-exportLogging-Functionality.R:18:5
• Tests to be developed for 3.0.0 release (9):
  test-151-exportRecords-ArgumentValidation.R:6:5,
  test-151-exportRecords-Functionality.R:6:5,
  test-152-exportRecordsWithDags.R:6:5,
  test-153-exportRecordsWithEvents.R:6:5,
  test-155-exportRecordsWithRepeatingInstruments.R:6:5,
  test-180-exportRecordsOffline-ArgumentValidation.R:6:5,
  test-180-exportRecordsOffline-Functionality.R:6:5,
  test-190-exportReports-ArgumentValidation.R:6:5,
  test-190-exportReports-Functionality.R:6:5

[ FAIL 0 | WARN 0 | SKIP 10 | PASS 2065 ]
```

**Comment by spgarbet** [comment]
8v76d6

**Comment by obregos** [comment]
```
══ Results ═════════════════════
Duration: 668.2 s

── Skipped tests (13) ─────────
• No logs for test window in past (1): test-800-exportLogging-Functionality.R:18:5
• No value provided for EXPORT_REPORTS_ID. Tests Skipped (3): test-240-exportTypedReports-Functionality.R:8:5, test-240-exportTypedReports-Functionality.R:54:5,
  test-240-exportTypedReports-Functionality.R:72:5
• Tests to be developed for 3.0.0 release (9): test-151-exportRecords-ArgumentValidation.R:6:5, test-151-exportRecords-Functionality.R:6:5, test-152-exportRecordsWithDags.R:6:5,
  test-153-exportRecordsWithEvents.R:6:5, test-155-exportRecordsWithRepeatingInstruments.R:6:5, test-180-exportRecordsOffline-ArgumentValidation.R:6:5,
  test-180-exportRecordsOffline-Functionality.R:6:5, test-190-exportReports-ArgumentValidation.R:6:5, test-190-exportReports-Functionality.R:6:5

[ FAIL 0 | WARN 0 | SKIP 13 | PASS 2062 ]
```

**Comment by spgarbet** [comment]
```
── R CMD check results ─────────────────────────────── redcapAPI 2.9.3 ────
Duration: 44.4s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔

R CMD check succeeded
```

**Comment by jubilee2** [comment]
── R CMD check results ────────────────────────────────────────────────────────────────────── redcapAPI 2.9.3 ────
Duration: 27.2s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔

R CMD check succeeded

**Comment by jubilee2** [comment]
── Skipped tests (14) ────────────────────────────────────────────────────────────────────────────────────────────
• No Data Quality Project provided. Tests Skipped (1): test-904-exportDataQuality.R:6:3
• No logs for test window in past (1): test-800-exportLogging-Functionality.R:18:5
• No value provided for EXPORT_REPORTS_ID. Tests Skipped (3): test-240-exportTypedReports-Functionality.R:8:5,
  test-240-exportTypedReports-Functionality.R:54:5, test-240-exportTypedReports-Functionality.R:72:5
• Tests to be developed for 3.0.0 release (9): test-151-exportRecords-ArgumentValidation.R:6:5,
  test-151-exportRecords-Functionality.R:6:5, test-152-exportRecordsWithDags.R:6:5,
  test-153-exportRecordsWithEvents.R:6:5, test-155-exportRecordsWithRepeatingInstruments.R:6:5,
  test-180-exportRecordsOffline-ArgumentValidation.R:6:5, test-180-exportRecordsOffline-Functionality.R:6:5,
  test-190-exportReports-ArgumentValidation.R:6:5, test-190-exportReports-Functionality.R:6:5

[ FAIL 0 | WARN 0 | SKIP 14 | PASS 2060 ]

**Review by jubilee2 — APPROVED** [review]

**Pull request merged**